src/Interactive/EditorFeatures/Core/Extensibility/Interactive/ResetInteractive.cs

```diff
@@ -0,0 +1,134 @@
+                    namespacesToImport.ToImmutableArray(),
+                    projectDirectory,
+                    waitContext);
+
+                // Once we're done resetting, dismiss the wait indicator and focus the REPL window.
+                return resetInteractiveTask.SafeContinueWith(
+                    _ =>
+                    {
+                        waitContext.Dispose();
+                        ExecutionCompleted?.Invoke(this, new EventArgs());
+                    },
+                    TaskScheduler.FromCurrentSynchronizationContext());
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private async Task ResetInteractiveAsync(
+            IInteractiveWindow interactiveWindow,
+            ImmutableArray<string> referencePaths,
+            ImmutableArray<string> referenceSearchPaths,
+            ImmutableArray<string> sourceSearchPaths,
+            ImmutableArray<string> namespacesToImport,
+            string projectDirectory,
+            IWaitContext waitContext)
+        {
+            // First, open the repl window.
+            IInteractiveEvaluator engine = interactiveWindow.Evaluator;
+
+            // If the user hits the cancel button on the wait indicator, then we want to stop the
+            // build.
+            waitContext.CancellationToken.Register(() =>
+                CancelBuildProject(), useSynchronizationContext: true);
+
+            // First, start a build.
+            // If the build fails do not reset the REPL.
+            var builtSuccessfully = await BuildProject().ConfigureAwait(true);
+            if (!builtSuccessfully)
+            {
+                return;
+            }
+
+            // Then reset the REPL
+            waitContext.Message = InteractiveEditorFeaturesResources.ResettingInteractive;
+            await interactiveWindow.Operations.ResetAsync(initialize: true).ConfigureAwait(true);
+
+            // TODO: load context from an rsp file.
+
+            // Now send the reference paths we've collected to the repl.
+            // The SetPathsAsync method is not available through an Interface.
+            // Execute the method only if the cast to a concrete InteractiveEvaluator succeeds.
+            InteractiveEvaluator interactiveEngine = engine as InteractiveEvaluator;
+            if (interactiveEngine != null)
+            {
+                await interactiveEngine.SetPathsAsync(referenceSearchPaths, sourceSearchPaths, projectDirectory).ConfigureAwait(true);
+            }
+
+            await interactiveWindow.SubmitAsync(new[]
+            {
+                referencePaths.Select(_createReference).Join("\r\n"),
+                namespacesToImport.Select(_createImport).Join("\r\n")
+            }).ConfigureAwait(true);
+        }
+
+        /// <summary>
+        /// Gets the properties of the currently selected projects necessary for reset.
+        /// </summary>
+        protected abstract bool GetProjectProperties(
```

Comment:
Consider having this return `ImmutableArray`, rather than `List`, to make the interface cleaner.